fcm/fcm.go

```diff
@@ -0,0 +1,216 @@
+package fcm
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+import (
+	"github.com/google/cloud-print-connector/log"
+	"github.com/google/cloud-print-connector/notification"
+)
+
+const (
+	gcpFcmSubscribePath = "fcm/subscribe"
+)
+
+type FCM struct {
+	fcmServerBindURL string
+	cachedToken      string
+	tokenRefreshTime time.Time
+	clientID         string
+	proxyName        string
+	fcmTTLSecs       float64
+	FcmSubscribe     func(string) (interface{}, error)
+
+	notifications chan<- notification.PrinterNotification
+	dead          chan struct{}
+
+	quit chan struct{}
+}
+
+type Notification struct {
+	Data   [][]interface{}
+	Number int
+}
+
+type Data struct {
+	Notification string `json:"notification"`
+	Subtype      string `json:"subtype"`
+}
+
+type FcmMessage struct {
+	Category    string `json:"category"`
+	CollapseKey string `json:"collapse_key"`
+	Data        Data   `json:"data"`
+	From        string `json:"from"`
+	MessageID   string `json:"message_id"`
+	TimeToLive  int    `json:"time_to_live"`
+}
+
+func NewFCM(clientID string, proxyName string, fcmServerBindURL string, FcmSubscribe func(string) (interface{}, error), notifications chan<- notification.PrinterNotification) (*FCM, error) {
+	f := FCM{
+		fcmServerBindURL,
+		"",
+		time.Time{},
+		clientID,
+		proxyName,
+		0,
+		FcmSubscribe,
+		notifications,
+		make(chan struct{}),
+		make(chan struct{}),
+	}
+	return &f, nil
+}
+
+//  get token from GCP and connect to FCM.
+func (f *FCM) Init() {
+	iidToken := f.GetTokenWithRetry()
+	if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+		for err != nil {
+			log.Errorf("FCM restart failed, will try again in 10s: %s", err)
+			time.Sleep(10 * time.Second)
+			err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
+		}
+		log.Info("FCM conversation restarted successfully")
+	}
+
+	go f.KeepFcmAlive()
+}
+
+// Quit terminates the FCM conversation so that new jobs stop arriving.
+func (f *FCM) Quit() {
+	// Signal to KeepFCMAlive.
+	close(f.quit)
+}
+
+// Fcm notification listener
+func (f *FCM) ConnectToFcm(fcmNotifications chan<- notification.PrinterNotification, iidToken string, dead chan<- struct{}, quit chan<- struct{}) error {
+	log.Debugf("Connecting to %s?token=%s", f.fcmServerBindURL, iidToken)
+	resp, err := http.Get(fmt.Sprintf("%s?token=%s", f.fcmServerBindURL, iidToken))
+	if err != nil {
+		// failed for ever no need to retry
+		quit <- struct{}{}
+		return err
+	}
+	if resp.StatusCode == 200 {
+		reader := bufio.NewReader(resp.Body)
+		go func() {
+			for {
+				line, err := reader.ReadBytes('\n')
```

Comment:
does reader.ReadBytes('\n') read until it gets a new line? If so, wouldn't messages be delayed since they don't end with a new line? In the examples below, you seem to get a number, then new line, then that previous number of bytes. Shouldn't the receiving side follow that?

Reply:
Fixed.
This is a very good find.